### ISSUE TITLE
Frames should be Serializable

### ISSUE THREAD

**Issue by guidou** [alvestrand/webrtc-media-streams#5]
In order to support running streams on a Worker, chunks must be marked serializable, since that is the mechanism streams use to send chunks to a Worker. 
For simplicity and efficiency, we should neuter frames once they're serialized, so that it is easier to make the deserialized frame on the Worker side reuse the underlying WebRTC frames.

**Comment by guidou** [comment]
cc @alvestrand

**Comment by guest271314** [comment]
@guidou What is the underlying data of `data` and `additionalData`? Given 

```
RTCEncodedVideoFrame {type: "key", timestamp: 0, data: ArrayBuffer(5826), additionalData: ArrayBuffer(5)}
additionalData: ArrayBuffer(5) {}
```

how to create an `ImageBitmap` or `ImageData` from the video frame?

What exactly do `data` and `additionalData` comprise? What is being serialized?

**Comment by guidou** [comment]
data is the encoded frame, additionalData contains metadata in a format exposed by the webrtc library. In a future iteration we'll replace additionalData with structured metadata.

**Comment by guest271314** [comment]
Kindly illuminate, when you have the time, . A WebRTC specific frame? A VP8 or other codec frame? What is the algorithm to decode the frame to be used as a displayed image, e.g, `ImageBitmap`; `ImageData`; WebP image, and similarly for audio, in JavaScript?

Is the concept to modify or pass-through encoded frames to be output from `WritableStream` or `ReadableStream` to `MediaStreamTrack`?

**Comment by guidou** [comment]
Spec wise, it's a frame encoded with whatever codec is being used. In the current Chromium implementation (not yet shipped) we have tested with VP8.
The concept is to read the data from the readable stream, process it (including analyzing/changing/replacing data),  and writing back to the writable stream to put the frame back in the WebRTC stack.

**Comment by guest271314** [comment]
How to decode the frame? What is the algorithm? And conversely, given an array of images (e.g., all images in a video, e.g., https://plnkr.co/edit/k6Itrjzzj60vT2fC), how to encode the individual frame and insert the array of encoded frames (including `Float32Array` for audio to encoded frame (Opus?), https://github.com/thenickdude/webm-writer-js/issues/1) into a `MediaStreamTrack` without using `<canvas>` or `HTMLMediaElement`? 

How is this envisioned to connect with Web Audio API?

**Comment by guest271314** [comment]
That is, the use case of streaming images https://plnkr.co/edit/4Tb91b?preview and audio (https://github.com/WebAudio/web-audio-api-v2/issues/61; https://plnkr.co/edit/nECtUZ?preview) from `Worker`: How to encode a `Uint8Array` to the WebRTC frame output by serialization to a `MediaStreamTrack` from `Worker` (or `Worklet`) to `MediaStreamTrack`?

**Comment by guidou** [comment]
> How to decode the frame? What is the algorithm? And conversely, given an array of images (e.g., all images in a video, e.g., https://plnkr.co/edit/k6Itrjzzj60vT2fC), how to encode the individual frame and insert the array of encoded frames (including `Float32Array` for audio to encoded frame (Opus?), [thenickdude/webm-writer-js#1](https://github.com/thenickdude/webm-writer-js/issues/1)) into a `MediaStreamTrack` without using `<canvas>` or `HTMLMediaElement`?
>

The frame is decoded by the webrtc stack once you write it back to the writable stream on the receiver side. At this time, this API is not intended to support integration with external codecs, but with the internal codecs of the webrtc stack. 
 
> How is this envisioned to connect with Web Audio API?
This is not envisioned to connect with Web Audio at the moment.

**Comment by guest271314** [comment]
Where in Chromium source code can we locate the data structure of a WebRTC encoded frame and the algorithm to decode the frame (e.g., VP8, VP9, Opus)?

> The frame is decoded by the webrtc stack once you write it back to the writable stream on the receiver side.

If the frame is serializable and

> (including analyzing/changing/replacing data)

is part of the use case, how to perform the above use case tasks? 

For example, inverting the colors of a VP8 frame, converting YUV to RGB and drawing a specific image onto a canvas from the serialized frame, and grabbing an Opus frame, converting to `Float32Array` and passing that data to `AudioWorklet`?

What is the data structure of a WebRTC frame? Is there a diagram of what an WebRTC frame "looks like" in a "human readable" form?

**Comment by fippo** [comment]
This is an encoded frame. You can not change the colors of an encoded frame easily.

The relevant bitstream format for VP8 is described in https://tools.ietf.org/html/rfc6386#section-9.1

**Comment by guest271314** [comment]
@fippo Am currently using `canvas.toBlob()` to get all images from a video https://plnkr.co/edit/k6Itrjzzj60vT2fC?preview (can be thousands of images, depending) which am able to save locally, load and play back as a video at `canvas`. Not simple to synchronize with audio stored as `Float32Array` when streaming back. Not really possible to easily stream a JavaScript object (`JSON`), therefore after running the basic tests above, am now considering testing substiuting encoded frams for WebP images and `Float32Array`. Just want to make sure that `additionalData` is needed or not. If not can save the `Uint8Array`s only as individual files. 

Is the encoded frame exactly and only a VP8 bitstream, without any WebRTC specific attached data? 

If the encoded frame can actually be decoded it should be possible to decode the frame by piping to a `WritableStream`, paint the frame onto a `canvas`, change color, re-write frame to a `ReadableStream`, store new frame?

For now just want to make sure `data` can be stored and re-used later, and if `additionalData` is necessary as part of the decoding process.

It should be possible to write the stored frames to a `PeerConnection` later, correct?

**Comment by alvestrand** [comment]
@guest271314 there are books available to learn the basics of video encoding.
You could probably do worse than starting with https://www.amazon.com/Video-Codec-Design-Developing-Compression/dp/0471485535/ - it gets good reviews.
TL;DR: An easily manipulatable form of a frame such as what we put on a canvas is several megabytes. A frame as compressed into an "encoded frame" is on the order of kilobytes. The software to do the compression / decompression (which is lossy, btw, but the compression is designed to minimize the human-visible impact of the information loss) is rather complex (tens of thousands of lines of code).
Stored frames are dependent on each other - you have to start off with a keyframe. So storing one frame and reusing it later is not possible.

**Comment by guest271314** [comment]
> So storing one frame and reusing it later is not possible.

Why not? The keyframe can be re-written in whichever order once the user has the serialized raw data, correct? 

If 300x200 video is captured, then 640x480 video is captured and both serialized, are you stating it would be impossible to wait a day, then stream 640x480 then 300x200, then capture live video at 800x600 and add that video to the stream? Given a function is composed which re-orders keyframes; similar to how `ts-ebml` is able to set the duration on WebM files produced by Chromium implementation of `MediaRecorder`, that has an initial duration of `Infinity`?

Should this issue https://github.com/w3c/webrtc-pc/issues/2507 be filed in this repository?

**Comment by guest271314** [comment]
We should be able to do with serialized WebRTC frames what is possible with webm-writer without the container https://plnkr.co/edit/Inb676?preview.
